include/ginkgo/core/solver/solver_base.hpp

```diff
@@ -48,6 +50,27 @@ namespace gko {
 namespace solver {
 
 
+/**
+ * Traits class providing information on the type and location of workspace
+ * vectors inside a solver.
+ */
+template <typename Solver>
+struct solver_workspace_traits {
+    // number of vectors used by this workspace
+    static int num_vectors(const Solver&) { return 0; }
```

Comment:
It would be nice to have a compile-time check that all vector (and array) ids are actually less than `num_vectors`, but I think that requires C++20...  And that all ids are unique...